inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/config/DataProviderCallScheduler.java

```diff
@@ -0,0 +1,158 @@
+package rocks.inspectit.oce.core.instrumentation.config;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Value;
+import lombok.val;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.stereotype.Component;
+import rocks.inspectit.oce.core.config.model.instrumentation.dataproviders.DataProviderCallSettings;
+import rocks.inspectit.oce.core.instrumentation.config.model.DataProviderCallConfig;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class encapsulates the logic how data providers are ordered.
+ * The approach is that first a dependency graph is generated.
+ * In this dependency graph the nodes are the data-keys and the data provider calls writing them,
+ * the edges indicate a dependency between data keys and the corresponding data provider calls.
+ * The edges are direct and and edge from data_A to data_B reads as
+ * "The provider writing data_B must be executed before the provider of data_A"
+ */
+@Component
+public class DataProviderCallScheduler {
+
+    public List<Pair<String, DataProviderCallConfig>> orderDataProviderCalls(Map<String, DataProviderCallConfig> calls) throws CyclicDataDependencyException {
+        Set<String> dataKeys = calls.keySet();
+        Map<String, Set<String>> dependencyGraph = new HashMap<>();
+        calls.forEach((dataKey, providerCall) -> collectDependencies(dataKey, providerCall, dependencyGraph));
+
+        //filter out data keys which are not written
+        val filteredDependencyGraph = dependencyGraph.entrySet().stream()
+                .filter(e -> dataKeys.contains(e.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        return getInTopologicalOrder(filteredDependencyGraph).stream()
+                .map(data -> Pair.of(data, calls.get(data)))
+                .collect(Collectors.toList());
+    }
+
+    private void collectDependencies(String dataKey, DataProviderCallConfig providerCall, Map<String, Set<String>> dependencyGraph) {
+        DataProviderCallSettings callSettings = providerCall.getCallSettings();
+
+        Set<String> afterDependencies = new HashSet<>();
+        //this data provider uses the following data keys as input arguments, therefore depends on them
+        val dataInputs = callSettings.getDataInput().values();
+        afterDependencies.addAll(dataInputs);
```

Comment:
For me, `afterDependencies` and `beforeDependencies` was quite confusing at the beginning because `beforeDependencies` are dps (data providers) which are called after the actual one, right?

Can we rename those to (I have no good idea), for example:
`afterDependencies` => `dependencies` or `preconditions`
`beforeDependencies` => `followupDataProviders`, `followupDependencies` (I dont know if this makes sense :D )

Reply:
How about the following names:
afterDependencies: executedAfter
beforeDependencies: executedBefore
?

Reply:
Passt!